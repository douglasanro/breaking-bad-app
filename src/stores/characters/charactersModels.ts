import ICharacter from 'models/ICharacter';

export type TCharacterFilterStatus = 'Alive' | 'Dead' | '';

export type TFilterTypes = 'status';

export interface ICharacterFilter {
  status: TCharacterFilterStatus;
};

export interface ICharacterState {
  list: ICharacter[];
  hasNextPage: boolean;
  filter: ICharacterFilter;
  searchTerm: string;
};

export interface IFetchCharacters {
  limit: number;
  offset?: number;
};
