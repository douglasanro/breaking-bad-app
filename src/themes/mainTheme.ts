import { unstable_createMuiStrictModeTheme as createMuiTheme } from '@material-ui/core';

const mainTheme = createMuiTheme({
  palette: {
    primary: {
      light: '#69A704',
      main: '#3A7116',
      dark: '#064500',
      contrastText: '#FFF',
    },
    secondary: {
      light: '#FFFF51',
      main: '#DFD400',
      dark: '#A9A300',
      contrastText: '#000',
    },
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        "*, *::before, *::after": {
          boxSizing: "content-box",
        },
        body: {
          background: '#FFF'
        },
      },
    },
  },
});

export default mainTheme;
