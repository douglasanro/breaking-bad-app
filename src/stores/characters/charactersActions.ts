import { action, createAsyncAction } from 'typesafe-actions';
import { ICharacterFilter } from './charactersModels';
import ICharacter from 'models/ICharacter';

export enum ECharactersActions {
  FETCH_CHARACTERS_REQUEST = '@characters/FETCH_CHARACTERS_REQUEST',
  FETCH_CHARACTERS_SUCCESS = '@characters/FETCH_CHARACTERS_SUCCESS',
  FETCH_CHARACTERS_FAILURE = '@characters/FETCH_CHARACTERS_FAILURE',
  CLEAR_CHARACTERS = '@characters/CLEAR_CHARACTERS',
  HAS_NEXT_PAGE = '@characters/HAS_NEXT_PAGE',
  SET_SEARCH_TERM = '@characters/SET_SEARCH_TERM',
  SET_FILTER = '@characters/SET_FILTER'
}

export const setCharacters = createAsyncAction(
  ECharactersActions.FETCH_CHARACTERS_REQUEST,
  ECharactersActions.FETCH_CHARACTERS_SUCCESS,
  ECharactersActions.FETCH_CHARACTERS_FAILURE,
)<undefined, ICharacter[], Error>();

export const clearCharacters = () => action(ECharactersActions.CLEAR_CHARACTERS);

export const setHasNextPage = (flag: boolean) => action(ECharactersActions.HAS_NEXT_PAGE, flag);

export const setSearchTerm = (term: string) => action(ECharactersActions.SET_SEARCH_TERM, term);

export const setFilter = ({ status }: ICharacterFilter) => action(ECharactersActions.SET_FILTER, { status });
