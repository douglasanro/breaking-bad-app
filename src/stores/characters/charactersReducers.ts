import { createReducer } from 'typesafe-actions';
import { ECharactersActions } from './charactersActions';
import { ICharacterState } from './charactersModels'

export const charactersInitialState: ICharacterState = {
  list: [],
  hasNextPage: true,
  filter: {
    status: ''
  },
  searchTerm: ''
};

const charactersReducers = createReducer(charactersInitialState, {
  [ECharactersActions.FETCH_CHARACTERS_SUCCESS]: (state, action) => ({
    ...state,
    list: [...new Set([...state.list, ...action.payload])]
  }),
  [ECharactersActions.CLEAR_CHARACTERS]: (state) => ({
    ...state,
    list: []
  }),
  [ECharactersActions.HAS_NEXT_PAGE]: (state, action) => ({
    ...state,
    hasNextPage: action.payload
  }),
  [ECharactersActions.SET_SEARCH_TERM]: (state, action) => ({
    ...state,
    searchTerm: action.payload
  }),
  [ECharactersActions.SET_FILTER]: (state, action) => ({
    ...state,
    filter: {...state.filter, ...action.payload}
  }),
});

export default charactersReducers;
