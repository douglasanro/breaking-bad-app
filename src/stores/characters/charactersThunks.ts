import { Dispatch } from 'redux';
import axios from 'axios';
import { setCharacters, setHasNextPage } from './charactersActions';
import { IFetchCharacters } from './charactersModels';
import ICharacter from 'models/ICharacter';
import { rootState } from 'stores/rootStore';

const baseUrl = 'https://breakingbadapi.com/api';
const endpoint = '/characters';
const defaultOffset = 0;

export const getCharacters = ({ limit, offset = defaultOffset }: IFetchCharacters) => async ( dispatch: Dispatch, getState: () => rootState) => {
  const { characters } = getState();
  const { searchTerm } = characters;
  try {
    dispatch(setCharacters.request());
    const searchByName = searchTerm ? `&name=${searchTerm}` : '';
    const { data } = await axios.get(`${baseUrl}${endpoint}?limit=${limit}&offset=${offset}${searchByName}`);

    if(!data.length) {
      dispatch(setHasNextPage(false));
      return;
    }

    const effectData = data.map(({ status, ...otherArgs}: ICharacter) => {
      const currentStatus = ['Alive', 'Dead'][Number(status !== 'Alive')];

      return {...otherArgs, status: currentStatus };
    })

    dispatch(setHasNextPage(true));
    dispatch(setCharacters.success(effectData));
  } catch (error) {
    dispatch(setCharacters.failure(error));
  }
};
