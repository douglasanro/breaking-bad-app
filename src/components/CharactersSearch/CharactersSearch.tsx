import React from 'react';
import { useDispatch } from 'react-redux';
import debounce from 'lodash/debounce';
import InputAdornment from '@material-ui/core/InputAdornment';
import TextField from '@material-ui/core/TextField';
import SearchIcon from '@material-ui/icons/Search';
import { clearCharacters, setSearchTerm } from 'stores/characters/charactersActions';
import { getCharacters } from 'stores/characters/charactersThunks';

const CharactersSearch: React.FC = () => {
  const dispatch = useDispatch();

  const searchCharacter = (event: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(clearCharacters());
    dispatch(setSearchTerm(event.target.value));

    if (event.target.value) {
      dispatch(getCharacters({ limit: 10 }));
    }
  };

  return (
    <TextField
      fullWidth
      label="Search character..."
      variant="outlined"
      onChange={debounce(searchCharacter, 500)}
      data-testid="charactersSearch"
      inputProps={{
        'data-testid': 'charactersSearchInput',
      }}
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <SearchIcon color="primary" />
          </InputAdornment>
        ),
      }}
    />
  );
};

export default CharactersSearch;
