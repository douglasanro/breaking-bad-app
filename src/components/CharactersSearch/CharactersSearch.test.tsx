import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from '@material-ui/core/styles';
import CharactersSearch from './CharactersSearch';
import { charactersInitialState } from 'stores/characters/charactersReducers';
import mainTheme from 'themes/mainTheme';

jest.mock('lodash/debounce', () => (fn: any) => {
  fn.cancel = jest.fn();
  return fn;
});

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
let store: ReturnType<typeof mockStore>;

const setupTest = () =>
  render(<CharactersSearch />, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </Provider>
    ),
  });

beforeEach(() => {
  store = mockStore({
    characters: charactersInitialState,
  });
});

test('Render CharactersSearch correctly', () => {
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});

test('Should search correctly', () => {
  const { getByTestId } = setupTest();
  const expectedAction = [
    {
      type: '@characters/CLEAR_CHARACTERS',
      payload: undefined,
      meta: undefined,
      error: undefined,
    },
    {
      type: '@characters/SET_SEARCH_TERM',
      payload: 'Test',
      meta: undefined,
      error: undefined,
    },
    { type: '@characters/FETCH_CHARACTERS_REQUEST' },
  ];
  fireEvent.change(getByTestId(/characters-search-input/i), { target: { value: 'Test' } });

  expect(store.getActions()).toEqual(expectedAction);
});

test('Should fill and clear search input correctly', () => {
  const { getByTestId } = setupTest();
  const expectedAction = [
    {
      type: '@characters/CLEAR_CHARACTERS',
      payload: undefined,
      meta: undefined,
      error: undefined,
    },
    {
      type: '@characters/SET_SEARCH_TERM',
      payload: 'Test',
      meta: undefined,
      error: undefined,
    },
    { type: '@characters/FETCH_CHARACTERS_REQUEST' },
    {
      type: '@characters/CLEAR_CHARACTERS',
      payload: undefined,
      meta: undefined,
      error: undefined,
    },
    {
      type: '@characters/SET_SEARCH_TERM',
      payload: '',
      meta: undefined,
      error: undefined,
    },
  ];
  fireEvent.change(getByTestId(/characters-search-input/i), { target: { value: 'Test' } });
  fireEvent.change(getByTestId(/characters-search-input/i), { target: { value: '' } });

  expect(store.getActions()).toEqual(expectedAction);
});
