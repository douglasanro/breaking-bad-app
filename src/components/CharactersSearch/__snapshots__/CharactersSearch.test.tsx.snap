// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CharactersSearch correctly 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink="false"
    >
      Search character...
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        data-testid="characters-search-input"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Search character...
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
